ext/spl/spl_rewindablekeyvalueiterator.c

```diff
@@ -0,0 +1,473 @@
+PHP_METHOD(RewindableKeyValueIterator, key)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	if (intern->offset >= intern->array.size) {
+		RETURN_NULL();
+	}
+	RETURN_COPY(&intern->array.entries[intern->offset].key);
+}
+
+/* Returns the value at the current offset */
+PHP_METHOD(RewindableKeyValueIterator, current)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	if (intern->offset >= intern->array.size) {
+		RETURN_NULL();
+	}
+	RETURN_COPY(&intern->array.entries[intern->offset].value);
+}
+
+/* Advance the iterator */
+PHP_METHOD(RewindableKeyValueIterator, next)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	if (intern->offset < intern->array.size) {
+		intern->offset++;
+	}
+}
+
+/* Rewind the iterator */
+PHP_METHOD(RewindableKeyValueIterator, rewind)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	intern->offset = 0;
+}
+
+/* Check if iterator is valid */
+PHP_METHOD(RewindableKeyValueIterator, valid)
+{
+	spl_rewindablekeyvalueiterator_object  *intern;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	intern = Z_REWINDABLEKEYVALUEITERATOR_P(ZEND_THIS);
+	RETVAL_BOOL(intern->offset < intern->array.size);
+}
+
+// TODO: Implement internal iterator functions for better performance? See spl_fixedarray.c
+
+PHP_MINIT_FUNCTION(spl_rewindablekeyvalueiterator)
+{
+	REGISTER_SPL_STD_CLASS_EX(RewindableKeyValueIterator, spl_rewindablekeyvalueiterator_new, class_RewindableKeyValueIterator_methods);
+	memcpy(&spl_handler_RewindableKeyValueIterator, &std_object_handlers, sizeof(zend_object_handlers));
+
+	spl_handler_RewindableKeyValueIterator.offset          = XtOffsetOf(spl_rewindablekeyvalueiterator_object, std);
+	spl_handler_RewindableKeyValueIterator.clone_obj       = spl_rewindablekeyvalueiterator_object_clone;
+	spl_handler_RewindableKeyValueIterator.count_elements  = spl_rewindablekeyvalueiterator_object_count_elements;
+	spl_handler_RewindableKeyValueIterator.get_properties  = spl_rewindablekeyvalueiterator_object_get_properties;
+	spl_handler_RewindableKeyValueIterator.get_gc          = spl_rewindablekeyvalueiterator_object_get_gc;
+	spl_handler_RewindableKeyValueIterator.dtor_obj        = zend_objects_destroy_object;
+	spl_handler_RewindableKeyValueIterator.free_obj        = spl_rewindablekeyvalueiterator_object_free_storage;
+
+	REGISTER_SPL_IMPLEMENTS(RewindableKeyValueIterator, Iterator);
+
+	spl_ce_RewindableKeyValueIterator->ce_flags |= ZEND_ACC_FINAL;
```

Comment:
You should include `ZEND_ACC_NO_DYNAMIC_PROPERTIES` as well.

Reply:
Done